src/connection.c

```diff
@@ -53,17 +53,16 @@ static status_e get_connection( struct service *sp, connection_s *cp )
       } else {
          cp->co_descriptor = accept( SVC_FD( sp ), &(cp->co_remote_address.sa),
                                      &sin_len ) ;
-	 if (cp->co_descriptor != -1)
-             M_SET( cp->co_flags, COF_NEW_DESCRIPTOR ) ;
-      }
-
-      if ( cp->co_descriptor == -1 )
-      {
-	 if ((errno == EMFILE) || (errno == ENFILE))
-	     cps_service_stop(sp, "no available descriptors");
-	 else
-             msg( LOG_ERR, func, "service %s, accept: %m", SVC_ID( sp ) ) ;
-         return( FAILED ) ;
+         if ( cp->co_descriptor == -1 )
+         {
+	    if ((errno == EMFILE) || (errno == ENFILE))
+	        cps_service_stop(sp, "no available descriptors");
+	    else
+                msg( LOG_ERR, func, "service %s, accept: %m", SVC_ID( sp ) ) ;
+            return( FAILED ) ;
+         }
+         M_SET( cp->co_flags, COF_NEW_DESCRIPTOR ) ;
+         M_SET( cp->co_flags, COF_HAVE_ADDRESS ) ;
```

Comment:
Should the address be set in the if( SC_WAITS( scp ) )?

It seems that in the other case it would not set the flags if we already havd the co_descriptor before

Reply:
On entry to get_connection() SVC_FD(sp) is the service's listening socket that receives client connections and there's a new connection in its backlog ready to be accept()'ed. The accept() will return a connected socket descriptor and store the client address in co_remote_address.

So when SC_WAITS()==0 (wait=no) get_connection() calls accept(), puts the connected socket descriptor into co_descriptor to pass to the service, and sets COF_HAVE_ADDRESS to indicate that co_remote_address is valid.

But when SC_WAITS()==1 (wait=yes) get_connection() doesn't do the accept(). It just copies the listening socket into co_descriptor to pass it to the service and the service does the accept(). So the client address remains unknown and COF_HAVE_ADDRESS should not be set.